

def initiallog():
    print("\n#######################################################################\n")
    print("     (^).-.(^)      ██████╗  █████╗ ███╗   ███╗███████╗███████╗")
    print("      / o_o \       ██╔══██╗██╔══██╗████╗ ████║██╔════╝██╔════╝")
    print("    __\(`-´)/__     ██████╔╝███████║██╔████╔██║███████╗█████╗ ")
    print("   (__ /'-'\ __)    ██╔══██╗██╔══██║██║╚██╔╝██║╚════██║██╔══╝ ")
    print("      || o ||       ██████╔╝██║  ██║██║ ╚═╝ ██║███████║███████╗")
    print("     _/ '_' \_      ╚═════╝ ╚═╝  ╚═╝╚═╝     ╚═╝╚══════╝╚══════╝")
    print("     `-´   `-´      Bacterial AMplicon SEquence processing pipeline")
    print("    Version: 1.0    By @Antton Alberdi & @Ostaizka Aizpurua\n")
    print("#######################################################################\n")

def settingslog(input,projectdir,paramsfile,logfile,taxdb,primer_for,primer_rev,ampliconlength,minampliconlength,maxampliconlength,maxerrors,minoverlap,trim_for,trim_rev,maxreads,chimerafold,copythreshold,taxfilter,adaptor_for,adaptor_rev):
    print("autoASV is running with the following parameters:\n")
    print("PATHS")
    print("\tInput file: "+str(input))
    print("\tProject directory: "+str(projectdir))
    print("\tParameters file: "+str(paramsfile))
    print("\tLog file: "+str(logfile))
    print("\tTaxonomy database: "+str(taxdb)+"\n")
    print("PRIMERS")
    print("\tForward primer: "+str(primer_for))
    print("\tReverse primer: "+str(primer_rev)+"\n")
    print("PARAMETERS")
    print("\tAverage amplicon length: "+str(ampliconlength))
    print("\tMaximum amplicon length: "+str(maxampliconlength))
    print("\tMinimum amplicon length: "+str(minampliconlength))
    print("\tMaximum Expected Errors: "+str(maxerrors))
    print("\tMinimum read overlap: "+str(minoverlap))
    print("\tForward read trimming length: "+str(trim_for))
    print("\tReverse read trimming length: "+str(trim_rev))
    print("\tMaximum number of reads for parameter optimisation: "+str(maxreads))
    print("\tChimera-assigning fold: "+str(chimerafold))
    print("\tRelative copy number threshold: "+str(copythreshold))
    print("\tTaxonomic filtering level: "+str(taxfilter)+"\n")
    print("\tForward adaptor sequence: "+str(adaptor_for))
    print("\tReverse adaptor sequence: "+str(adaptor_rev)+"\n")
    print("If you want to learn more about these parameters visit:")
    print("\thttps://github.com/anttonalberdi/autoasv\n")
    print("##########################################")
